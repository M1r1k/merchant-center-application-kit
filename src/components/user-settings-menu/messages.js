import { defineMessages } from 'react-intl';

export default defineMessages({
  userProfile: {
    id: 'UserSettingsMenu.userProfile',
    description: 'The label for user profile option',
    defaultMessage: 'My Profile',
  },
  logout: {
    id: 'UserSettingsMenu.logout',
    description: 'The label for logout option',
    defaultMessage: 'Logout',
  },
});
