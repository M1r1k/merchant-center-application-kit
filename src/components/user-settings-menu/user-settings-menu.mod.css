@import '@commercetools-local/ui-kit/materials/colors.mod.css';
@import '@commercetools-local/ui-kit/materials/spacings.mod.css';
@import '@commercetools-local/ui-kit/materials/shadows.mod.css';
@import '@commercetools-local/ui-kit/materials/borders.mod.css';

.container {
  position: relative;
}

.settings-container {
  cursor: pointer;
}

.menu {
  position: absolute;
  border: 1px var(--color-gray-60) solid;
  box-shadow: var(--shadow-7-first), var(--shadow-7-second);
  width: 200px;
  right: 0;
  top: 34px;
  padding: 0;
}

.menu::before {
  content: '';
  position: absolute;
  right: 30px;
  top: -6px;
  border-style: solid;
  border-width: 0 6px 6px;
  border-color: transparent transparent var(--color-white) transparent;
  z-index: 1;
}

.menu::after {
  content: '';
  position: absolute;
  right: 28px;
  top: -8px;
  border-style: solid;
  border-width: 0 8px 8px;
  border-color: transparent transparent var(--color-gray-60) transparent;
  z-index: 0;
}

.item {
  width: 100%;
  cursor: pointer;
  color: var(--color-black);
}

.item:hover {
  background-color: var(--color-gray-90);
}
