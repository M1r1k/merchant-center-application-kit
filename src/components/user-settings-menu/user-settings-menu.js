import PropTypes from 'prop-types';
import React from 'react';
import { Link } from 'react-router-dom';
import { FormattedMessage } from 'react-intl';
import Downshift from 'downshift';
import { ToggleFeature } from '@flopflip/react-broadcast';
import withMouseOverState from '@commercetools-local/ui-kit/hocs/with-mouse-over-state';
import { CaretDownIcon } from '@commercetools-local/ui-kit/icons';
import Text from '@commercetools-local/ui-kit/typography/text';
import Spacings from '@commercetools-local/ui-kit/materials/spacings';
import { LOGOUT_REASONS } from '@commercetools-local/constants';
import formatUserName from '@commercetools-local/utils/user';
import Avatar from '@commercetools-local/core/components/avatar';
import Card from '@commercetools-local/core/components/card';
import { MCSupportFormURL } from '../../constants';
import styles from './user-settings-menu.mod.css';
import messages from './messages';

export const UserAvatar = props => (
  <div onMouseOver={props.handleMouseOver} onMouseOut={props.handleMouseOut}>
    <Spacings.Inline alignItems="center">
      <div className={styles.avatar}>
        <Avatar
          email={props.email}
          firstName={props.firstName}
          lastName={props.lastName}
        />
      </div>
      <CaretDownIcon
        size="small"
        theme={props.isMouseOver ? 'grey' : 'black'}
      />
    </Spacings.Inline>
  </div>
);
UserAvatar.displayName = 'UserAvatar';
UserAvatar.propTypes = {
  email: PropTypes.string.isRequired,
  firstName: PropTypes.string,
  lastName: PropTypes.string,
  // Injected
  handleMouseOver: PropTypes.func.isRequired,
  handleMouseOut: PropTypes.func.isRequired,
  isMouseOver: PropTypes.bool.isRequired,
};
const UserAvatarWithHoverState = withMouseOverState(UserAvatar);

export default class UserSettingsMenu extends React.PureComponent {
  static displayName = 'UserSettingsMenu';

  static propTypes = {
    firstName: PropTypes.string,
    lastName: PropTypes.string,
    email: PropTypes.string.isRequired,
  };

  render() {
    return (
      <div className={styles.container} data-test="user-settings-menu">
        <Downshift
          render={({ isOpen, toggleMenu }) => (
            <div>
              <div
                className={styles['settings-container']}
                onClick={toggleMenu}
              >
                <UserAvatarWithHoverState
                  firstName={this.props.firstName}
                  lastName={this.props.lastName}
                  email={this.props.email}
                />
              </div>
              {isOpen && (
                <Card className={styles.menu}>
                  <Spacings.Inset scale="s">
                    <Text.Detail tone="secondary">
                      {formatUserName({
                        firstName: this.props.firstName,
                        lastName: this.props.lastName,
                      })}
                    </Text.Detail>
                  </Spacings.Inset>
                  <ToggleFeature flag="userProfile">
                    <Link to="/account/profile" onClick={toggleMenu}>
                      <div className={styles.item}>
                        <Spacings.Inset scale="s">
                          <FormattedMessage {...messages.userProfile} />
                        </Spacings.Inset>
                      </div>
                    </Link>
                  </ToggleFeature>
                  {
                    // FIXME: added as urgent request for GDPR changes
                    // Should be properly added on the CTP-1209 task
                  }
                  <a href={`https://commercetools.com/privacy`} target="_blank">
                    <div className={styles.item}>
                      <Spacings.Inset scale="s">
                        <FormattedMessage {...messages.privacyPolicy} />
                      </Spacings.Inset>
                    </div>
                  </a>
                  <a
                    href={MCSupportFormURL}
                    rel="noopener noreferrer"
                    target="_blank"
                    data-track-component="Navigation-Support-links"
                    data-track-event="click"
                    data-track-label="support_textlink"
                  >
                    <div className={styles.item}>
                      <Spacings.Inset scale="s">
                        <FormattedMessage {...messages.support} />
                      </Spacings.Inset>
                    </div>
                  </a>
                  <a
                    // NOTE: we want to redirect to a new page so that the
                    // server can remove things like cookie for access token.
                    href={`/logout?reason=${LOGOUT_REASONS.USER}`}
                    data-test="logout-button"
                  >
                    <div className={styles.item}>
                      <Spacings.Inset scale="s">
                        <FormattedMessage {...messages.logout} />
                      </Spacings.Inset>
                    </div>
                  </a>
                </Card>
              )}
            </div>
          )}
        />
      </div>
    );
  }
}
