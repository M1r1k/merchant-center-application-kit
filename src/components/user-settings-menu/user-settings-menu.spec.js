import React from 'react';
import { shallow } from 'enzyme';
import { UserSettingsMenu } from './user-settings-menu';

const createTestProps = props => ({
  firstName: 'John Test',
  lastName: 'Doe',
  email: 'john-doe@commercetools.de',
  handleMouseOver: jest.fn(),
  handleMouseOut: jest.fn(),
  isMouseOver: false,
  ...props,
});

describe('render', () => {
  let wrapper;
  let props;
  beforeEach(() => {
    props = createTestProps();
    wrapper = shallow(<UserSettingsMenu {...props} />);
  });

  it('outputs correct tree', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render `CaretDownIcon`', () => {
    expect(wrapper).toRender('CaretDownIcon');
  });

  it('should not render the menu', () => {
    expect(wrapper).not.toRender('Card');
  });
  it('should render `Avatar` component', () => {
    expect(wrapper).toRender('Avatar');
  });

  it('should pass prop firstName to `Avatar` component', () => {
    expect(wrapper.find('Avatar')).toHaveProp('firstName', 'John Test');
  });

  it('should pass prop lastName to `Avatar` component', () => {
    expect(wrapper.find('Avatar')).toHaveProp('lastName', 'Doe');
  });

  it('should pass prop email to `Avatar` component', () => {
    expect(wrapper.find('Avatar')).toHaveProp(
      'email',
      'john-doe@commercetools.de'
    );
  });
});

describe('callbacks', () => {
  describe('clicking to open the menu', () => {
    let wrapper;
    let props;
    beforeEach(() => {
      props = createTestProps();
      wrapper = shallow(<UserSettingsMenu {...props} />);
      wrapper.find({ className: 'settings-container' }).simulate('click');
    });
    it('outputs correct tree', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should render the menu', () => {
      expect(wrapper).toRender('Card');
    });
  });

  describe('when hovering the menu', () => {
    let props;
    let wrapper;
    beforeEach(() => {
      props = createTestProps({ isMouseOver: true });
      wrapper = shallow(<UserSettingsMenu {...props} />);
    });

    it('should set the grey theme for the icon on hovering', () => {
      expect(wrapper.find('CaretDownIcon')).toHaveProp('theme', 'grey');
    });
  });
});
