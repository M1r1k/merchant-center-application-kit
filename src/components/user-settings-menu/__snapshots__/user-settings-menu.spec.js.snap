// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`callbacks clicking to open the menu outputs correct tree 1`] = `
<div
  className="container"
  data-test="user-settings-menu"
>
  <div
    className="settings-container"
    onClick={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
  >
    <Inline
      alignItems="center"
      scale="s"
    >
      <div
        className="avatar"
      >
        <Avatar
          email="john-doe@commercetools.de"
          firstName="John Test"
          lastName="Doe"
          size="small"
        />
      </div>
      <CaretDownIcon
        size="small"
        theme="black"
      />
    </Inline>
  </div>
  <Card
    className="menu"
    type="raised"
  >
    <Inset
      scale="s"
    >
      <TextDetail
        tone="secondary"
      >
        John Test Doe
      </TextDetail>
    </Inset>
    <withFlagSubscription(withProps(FeatureToggled))
      flag="userProfile"
    >
      <Link
        replace={false}
        to="/profile"
      >
        <div
          className="item"
        >
          <Inset
            scale="s"
          >
            <FormattedMessage
              defaultMessage="My Profile"
              description="The label for user profile option"
              id="UserSettingsMenu.userProfile"
              values={Object {}}
            />
          </Inset>
        </div>
      </Link>
    </withFlagSubscription(withProps(FeatureToggled))>
    <Link
      data-test="logout-button"
      replace={false}
      to="/logout?reason=user"
    >
      <div
        className="item"
      >
        <Inset
          scale="s"
        >
          <FormattedMessage
            defaultMessage="Logout"
            description="The label for logout option"
            id="UserSettingsMenu.logout"
            values={Object {}}
          />
        </Inset>
      </div>
    </Link>
  </Card>
</div>
`;

exports[`render outputs correct tree 1`] = `
<div
  className="container"
  data-test="user-settings-menu"
>
  <div
    className="settings-container"
    onClick={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
  >
    <Inline
      alignItems="center"
      scale="s"
    >
      <div
        className="avatar"
      >
        <Avatar
          email="john-doe@commercetools.de"
          firstName="John Test"
          lastName="Doe"
          size="small"
        />
      </div>
      <CaretDownIcon
        size="small"
        theme="black"
      />
    </Inline>
  </div>
</div>
`;
